src/search/algorithms.clj

```diff
@@ -63,6 +63,35 @@
             new-visited (clojure.set/union children visited)]
         (recur new-frontier new-came-from new-visited)))))
 
+(defn a-star-search [children-fn cost-fn heuristic-fn start-state goal-state & {:keys [max-states] :or {max-states 1000000}}]
+  (loop [max-states max-states
+        frontier (pm/priority-map start-state 0)
+         came-from {}
+         cost-so-far {start-state 0}]
+    (if (or (neg? max-states)
+            (empty? frontier)
+            (>= (count came-from) max-states)
+            (= (first (peek frontier)) goal-state))
+      [came-from cost-so-far]
+      (let [current (first (peek frontier))
+            current-cost (cost-so-far current)
+            children (set (children-fn current))
+            children-costs (reduce #(assoc %1 %2 (+ current-cost (cost-fn current %2))) {} children)
+            children-to-add (filter #(or (not (contains? cost-so-far %))
+                                         (< (children-costs %) (cost-so-far %))) children)
+            heuristics (map (partial heuristic-fn goal-state) children-to-add)
```

Comment:
You never use this in `a-star-search`; it's pretty a copy/paste from `heuristic-search` that you never cleaned up.